crates/data_components/src/dynamodb/stream.rs

```diff
@@ -0,0 +1,617 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+use super::{
+    DowncastBuilderSnafu, FailedToAddItemToStructSnafu, FailedToCreateChangeBatchSnafu,
+    FailedToUnnestSnafu, Result, StreamError,
+};
+use crate::arrow::struct_builder::StructBuilder;
+use crate::cdc::{ChangeBatch, ChangeEnvelope, CommitChange, CommitError, changes_schema};
+use crate::dynamodb::arrow::append_item_to_struct_builder;
+use crate::dynamodb::unnest::unnest_dynamodb_item;
+use arrow::datatypes::Schema;
+use arrow_array::RecordBatch;
+use arrow_array::builder::{ArrayBuilder, ListBuilder, StringBuilder};
+use aws_sdk_dynamodb::types::AttributeValue as DynamoDbAttributeValue;
+use aws_sdk_dynamodbstreams::types::AttributeValue as StreamsAttributeValue;
+use aws_sdk_dynamodbstreams::types::OperationType;
+use dynamo_subscriber::error::StreamResult;
+use snafu::prelude::*;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+pub fn process_batch(
+    batch: StreamResult,
+    table_schema: &Arc<Schema>,
+    primary_keys: &[String],
+    unnest_depth: Option<usize>,
+    time_format: &str,
+) -> Result<ChangeEnvelope, StreamError> {
+    let changes_schema = changes_schema(table_schema).clone();
+
+    let batch = batch.map_err(|e| StreamError::FailedToReceiveMessage { source: e })?;
+
+    let mut changes_struct_builder =
+        StructBuilder::from_fields(changes_schema.fields().clone(), batch.len());
+
+    for record in batch {
+        let (op_str, item_data) = match (&record.event_name, &record.dynamodb) {
+            (Some(event_name), Some(dynamodb)) => match event_name {
+                OperationType::Insert | OperationType::Modify => {
+                    let Some(item) = &dynamodb.new_image else {
+                        continue;
+                    };
+                    let streams_item = streams_to_dynamodb_item(item.clone());
+
+                    let (unnested_streams_item, _) = match unnest_depth {
+                        None => (streams_item, HashSet::new()),
+                        Some(depth) => unnest_dynamodb_item(&streams_item, depth)
+                            .context(FailedToUnnestSnafu)?,
+                    };
+
+                    let op = if matches!(event_name, OperationType::Insert) {
+                        "c"
+                    } else {
+                        "u"
+                    };
+
+                    (op, unnested_streams_item)
+                }
+                OperationType::Remove => {
+                    let Some(keys_item) = &dynamodb.keys else {
+                        continue;
+                    };
+                    let streams_keys_item = streams_to_dynamodb_item(keys_item.clone());
+                    ("d", streams_keys_item)
+                }
+                operation => {
+                    tracing::debug!("Unexpected OperationType from DynamoDB Streams: {operation}",);
```

Comment:
Maybe higher than debug (If it occurs, we want to know regardless of if something is going wrong).